ext/spl/spl_heap.c

```diff
@@ -265,8 +293,23 @@ static void spl_ptr_heap_insert(spl_ptr_heap *heap, void *elem, void *cmp_userda
 		heap->max_size *= 2;
 	}
 
+	/* If we know this call came from non inherited SplPriorityQueue it's
+	 * possible to do specialization on the type of the priority parameter. */
+	if (pqueue && !Z_SPLHEAP_P(object)->fptr_cmp) {
+		int type = Z_TYPE(((spl_pqueue_elem*) elem)->priority);
+		spl_ptr_heap_cmp_func new_cmp =
+			(type == IS_LONG) ? spl_ptr_pqueue_elem_cmp_long :
+			((type == IS_DOUBLE) ? spl_ptr_pqueue_elem_cmp_double : spl_ptr_pqueue_elem_cmp);
+
+		if (heap->count == 0) { /* Specialize empty queue */
+			heap->cmp = new_cmp;
+		} else if (new_cmp != heap->cmp) { /* Despecialize on type conflict. */
+			heap->cmp = spl_ptr_pqueue_elem_cmp;
+		}
+	}
```

Comment:
Nifty idea. I think my preference would be to move this code up into SplPriorityQueue::insert(), as it's specific to that case.